sourmash/index.py

```diff
@@ -0,0 +1,123 @@
+"An Abstract Base Class for collections of signatures."
+
+from abc import ABCMeta, abstractmethod
+from collections import namedtuple
+
+# @CTB copied out of search.py to deal with import order issues, #willfix
+SearchResult = namedtuple('SearchResult',
+                          'similarity, match_sig, md5, filename, name')
+
+
+# compatible with Python 2 *and* 3:
+ABC = ABCMeta("ABC", (object,), {"__slots__": ()})
+
+
+class Index(ABC):
+    @abstractmethod
+    def find(self, search_fn, *args, **kwargs):
+        """ """
+
+    @abstractmethod
+    def search(self, signature, *args, **kwargs):
+        """ """
+
+    @abstractmethod
+    def gather(self, signature, *args, **kwargs):
+        """ """
+
+    @abstractmethod
+    def insert(self, node):
+        """ """
+
+    @abstractmethod
+    def save(self, path, storage=None, sparseness=0.0, structure_only=False):
+        """ """
+
+    @classmethod
+    @abstractmethod
+    def load(cls, location, leaf_loader=None, storage=None, print_version_warning=True):
+        """ """
+
+
+class LinearIndex(Index):
+    def __init__(self, signatures=[], filename=None):
```

Comment:
Noooooooooo... Don't set a default argument like that =]
http://effbot.org/zone/default-values.htm
```suggestion
    def __init__(self, signatures=None, filename=None):
        if signatures is None:
            signatures = []
```